### Datasets

https://data.cso.ie/table/SAP2022T1T2CTY

SAP2022T1T2CTY  - Population 2022 - Marital Status

And
https://data.cso.ie/table/SAP2016T1T2CTY

SAP2016T1T2CTY - Population by Sex and Marital Status 2016


## <font color='#1ABC9C'>MSC_DA_CA1<font color='#00A859'>
#### <font color='#'>**Programming for DA - Statistics for Data Analytics - Machine Learning for Data Analysis & Data Preparation & Visualisation
#### <font color='#1D76BA'>**Lecturer(s): Sam Weiss/ David Gonzalez, Bharathi Chakravarthi/ Marina Iantorno, Muhammad Iqbal and David McQuaid**
------
<font color='#E74C3C'>**Students name / ID** // Rosilene Francisca da Silva 2021090;

*This is a project using PYTHON programming language to use data to Analysis the Scenario of Population in Ireland.*</font>

In [1]:
# Importing the required libraries:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline

In [2]:
# Importing the Dataset 
datapop_2016 = pd.read_csv("2016population .csv")
datapop_2022 = pd.read_csv("2022population.csv")

Let's have a look at the information in our Datasets regarding the features it is a quickly preview the data and gain a basic understanding of its structure.

(Combine, the data sets contain more than 6,000 instances and 11 attributes + label output (quality)

In [3]:
datapop_2016.head()

,Statistic Label,Census Year,Sex,County,Marital Status,UNIT,VALUE
0,Population,2016,Males,Carlow,Single,Number,15971
1,Population,2016,Males,Carlow,Married,Number,10660
2,Population,2016,Males,Carlow,Separated,Number,696
3,Population,2016,Males,Carlow,Divorced,Number,557
4,Population,2016,Males,Carlow,Widowed,Number,581


In [4]:
 # Display the dimensions of the Dataset 1
datapop_2016.shape 

(558, 7)

In [5]:
datapop_2022.head()

,Statistic Label,Census Year,Administrative Counties 2019,Marital Status,Sex,UNIT,VALUE
0,Population,2022,Carlow County Council,Single,Males,Number,17173
1,Population,2022,Carlow County Council,Single,Females,Number,15769
2,Population,2022,Carlow County Council,Single,Both Sexes,Number,32942
3,Population,2022,Carlow County Council,Married,Males,Number,11488
4,Population,2022,Carlow County Council,Married,Females,Number,11492


In [6]:
# Display the dimensions of the Dataset 2
datapop_2022.shape 

(576, 7)

#### (When the databases are combined, they comprise almost 6,000 instances and 11 attributes.)

The databases have nearly identical numbers of attributes/features and instances/observations.There are 558 rows of population/2016 and 576 rows of Population/2022, each with 7 attributes.

#### Checkin null values in the datasets, we can do it in a variety of ways; I chose two to see whether there are any missing values in the datasets.

In [7]:
pop2016null = datapop_2016.isna().sum().sum()
pop2022null = datapop_2022.isna().sum().sum()
print ('Null values in datapop_2016: {}'.format(pop2016null))
print ('Null values in datapop_2022: {}'.format(pop2022null))

Null values in datapop_2016: 0
Null values in datapop_2022: 0


In [8]:
datapop_2016.isnull().values.any()
datapop_2022.isnull().values.any()

False

In [9]:
datapop_2016.isnull().sum()

Statistic Label    0
Census Year        0
Sex                0
County             0
Marital Status     0
UNIT               0
VALUE              0
dtype: int64

In [10]:
datapop_2022.isnull().sum()

Statistic Label                 0
Census Year                     0
Administrative Counties 2019    0
Marital Status                  0
Sex                             0
UNIT                            0
VALUE                           0
dtype: int64

The datasest does not contain any null values. As a result, there is no need to clean the datasets.

In [11]:
#Renaming some columns with the 'rename' methods because it's more appropriate before merging the datasets
datapop_2022 = datapop_2022.rename(columns={'Administrative Counties 2019': 'County',})
datapop_2022.head(2)

,Statistic Label,Census Year,County,Marital Status,Sex,UNIT,VALUE
0,Population,2022,Carlow County Council,Single,Males,Number,17173
1,Population,2022,Carlow County Council,Single,Females,Number,15769


In [12]:
datapop_2016.head()

,Statistic Label,Census Year,Sex,County,Marital Status,UNIT,VALUE
0,Population,2016,Males,Carlow,Single,Number,15971
1,Population,2016,Males,Carlow,Married,Number,10660
2,Population,2016,Males,Carlow,Separated,Number,696
3,Population,2016,Males,Carlow,Divorced,Number,557
4,Population,2016,Males,Carlow,Widowed,Number,581


### Merging the two datasets

In [66]:
# Concatenating the two Datasets
#frames = [datapop_2016, datapop_2022]
#data_pop = pd.concat(frames, ignore_index=True)
#data_pop.head()

In [73]:
# Perform the merge 'inner'
newdata_df = pd.merge(datapop_2016, datapop_2022, how ='inner', on=['County', 'Marital Status', 'Sex', 'UNIT', 'VALUE'])

# save the merged dataset to a new file if needed.
newdata_df.to_csv('newdata_df_dataset.csv', index=False)

#Merge by two or more different columns
merged_df = pd.merge(datapop_2016, datapop_2022, on = ['County', 'Administrative Counties 2019'])
merged_df

KeyError: 'Administrative Counties 2019'

In [72]:
datapop_2016 = pd.DataFrame({'Statistic Label': ['Population'],'Census Year': ['2016'], 'Sex': ['Males', 'Both Sexes', 'Females'],
                            'Marital Status': 'Single','Married', 'Separeted', 'Divorced', 'Widowed', 'Total'})
                             
datapop_2022 = pd.DataFrame({'Statistic Label': ['Population'],'Census Year': ['2022']})
  
# the default behaviour is join='outer'
# inner join
  
result = pd.concat([datapop_2016, datapop_2022], axis=1, join='inner')
display(result)

SyntaxError: ':' expected after dictionary key (73395135.py, line 2)

In [74]:
newdata_df.head(10)

,Statistic Label_x,Census Year_x,Sex,County,Marital Status,UNIT,VALUE,Statistic Label_y,Census Year_y


In [ ]:
#data_pop.isnull().values.any()

In [ ]:
#data_pop.isnull().sum()

In [86]:
# Merge based on index pandas
data_index = pd.concat([datapop_2016, datapop_2022])
data_index

,Statistic Label,Census Year,Sex,County,Marital Status,UNIT,VALUE
0,Population,2016,Males,Carlow,Single,Number,15971
1,Population,2016,Males,Carlow,Married,Number,10660
2,Population,2016,Males,Carlow,Separated,Number,696
3,Population,2016,Males,Carlow,Divorced,Number,557
4,Population,2016,Males,Carlow,Widowed,Number,581
...,...,...,...,...,...,...,...
571,Population,2022,Females,Ireland,Widowed,Number,155166
572,Population,2022,Both Sexes,Ireland,Widowed,Number,210083
573,Population,2022,Males,Ireland,Total,Number,2544549
574,Population,2022,Females,Ireland,Total,Number,2604590


I specifically selected this method to join these two datasets because they had different columns but the same common index. I integrated the data from both dataframes based on their shared index values by merging them on the index.

The datasets have have different column structures, In this case, concatenation might not be the best approach.

(https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html#pandas-dataframe-merge)

In [76]:
data_index.head(5)

,Statistic Label_x,Census Year_x,Sex_x,County_x,Marital Status_x,UNIT_x,VALUE_x,Statistic Label_y,Census Year_y,County_y,Marital Status_y,Sex_y,UNIT_y,VALUE_y
0,Population,2016,Males,Carlow,Single,Number,15971,Population,2022,Carlow County Council,Single,Males,Number,17173
1,Population,2016,Males,Carlow,Married,Number,10660,Population,2022,Carlow County Council,Single,Females,Number,15769
2,Population,2016,Males,Carlow,Separated,Number,696,Population,2022,Carlow County Council,Single,Both Sexes,Number,32942
3,Population,2016,Males,Carlow,Divorced,Number,557,Population,2022,Carlow County Council,Married,Males,Number,11488
4,Population,2016,Males,Carlow,Widowed,Number,581,Population,2022,Carlow County Council,Married,Females,Number,11492


In [77]:
data_index.isnull().values.any()

False

In [78]:
data_index.isnull()

,Statistic Label_x,Census Year_x,Sex_x,County_x,Marital Status_x,UNIT_x,VALUE_x,Statistic Label_y,Census Year_y,County_y,Marital Status_y,Sex_y,UNIT_y,VALUE_y
0,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
553,False,False,False,False,False,False,False,False,False,False,False,False,False,False
554,False,False,False,False,False,False,False,False,False,False,False,False,False,False
555,False,False,False,False,False,False,False,False,False,False,False,False,False,False
556,False,False,False,False,False,False,False,False,False,False,False,False,False,False


In [79]:
data_index.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 558 entries, 0 to 557
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Statistic Label_x  558 non-null    object
 1   Census Year_x      558 non-null    int64 
 2   Sex_x              558 non-null    object
 3   County_x           558 non-null    object
 4   Marital Status_x   558 non-null    object
 5   UNIT_x             558 non-null    object
 6   VALUE_x            558 non-null    int64 
 7   Statistic Label_y  558 non-null    object
 8   Census Year_y      558 non-null    int64 
 9   County_y           558 non-null    object
 10  Marital Status_y   558 non-null    object
 11  Sex_y              558 non-null    object
 12  UNIT_y             558 non-null    object
 13  VALUE_y            558 non-null    int64 
dtypes: int64(4), object(10)
memory usage: 65.4+ KB


#### The dataset has in total x rows each having x features.

In [80]:
data_index.dtypes

Statistic Label_x    object
Census Year_x         int64
Sex_x                object
County_x             object
Marital Status_x     object
UNIT_x               object
VALUE_x               int64
Statistic Label_y    object
Census Year_y         int64
County_y             object
Marital Status_y     object
Sex_y                object
UNIT_y               object
VALUE_y               int64
dtype: object

In [81]:
# Getting summary statistics
data_index.describe().transpose()

,count,mean,std,min,25%,50%,75%,max
Census Year_x,558.0,2016.000000,0.000000,2016.0,2016.00,2016.0,2016.0,2016.0
VALUE_x,558.0,34135.232975,55859.953742,332.0,2024.00,9813.0,43144.0,554554.0
Census Year_y,558.0,2022.000000,0.000000,2022.0,2022.00,2022.0,2022.0,2022.0
VALUE_y,558.0,36911.390681,59081.951257,407.0,2394.25,10921.5,47902.5,592713.0


In [82]:
#check for any duplicates 
duplicate = data_index[datapop_df.duplicated(keep='first')]
print(duplicate)

NameError: name 'datapop_df' is not defined

In [ ]:
Value = data_index['Census Year_x']
Value

In [ ]:
data_index.nunique()

In [ ]:
#Adding a Total Amout column
data_index['Total Amount'] = data_index.sum(axis=1)
data_index.head()

In [ ]:
#Renaming some columns with the 'rename' methods because it's more appropriate.
datapop_df = datapop_df.rename(columns={'Sex': 'Gender', 'VALUE': 'Value',})
datapop_df.head(2)

In [ ]:
value_nationality = datapop_df.groupby('Nationality')[['Value']].sum()
print(value_nationality.sort_values(by = ['Value']))

In [ ]:
plt.figure(figsize=(10,5))
sns.barplot(x='Socio Economic Group',y='Value',data=datapop_df,hue='Gender')
plt.xticks(rotation=70)
##Checking the distribution of Capital_gain and  Occupation as per gender##

In [ ]:
datapop_df['Census Year'].value_counts().plot.bar()

In [ ]:
sns.kdeplot(VALUE)

In [ ]:
#WE can use .T, right after sample(5) to transpose the Feature and this method will work in cascade and sequentially.
#datapop_df.head(5).T

In [ ]:
#If we wannt to change the colors
#We need to add the labels!

colors = ['red', 'yellow', 'violet']
fig, ax = plt.subplots()
ax.bar(value_nationality.index, value_nationality['Value'], color = colors, edgecolor = 'red')

plt.xlabel('Nationality')
plt.ylabel('Value')
plt.title('Nationalite Number')

In [ ]:
#visualizing data
sns.pairplot(datapop_df, hue = 'Socio Economic Group')